src/main/java/org/rumbledb/compiler/VariableDependenciesVisitor.java

```diff
@@ -0,0 +1,527 @@
+
+        visit(expression.getExpression(), null);
+        addInputVariableDependencies(expression, getInputVariableDependencies(expression.getExpression()));
+
+        removeInputVariableDependencies(
+            expression,
+            getOutputVariableDependencies(expression.getPreviousClause())
+        );
+        return null;
+    }
+
+    public Void visitGroupByClause(GroupByClause expression, Void argument) {
+        visit(expression.getPreviousClause(), null);
+        addOutputVariableDependencies(expression, getOutputVariableDependencies(expression.getPreviousClause()));
+
+        for (GroupByVariableDeclaration var : expression.getGroupVariables()) {
+            if (var.getExpression() != null) {
+                visit(var.getExpression(), null);
+                addInputVariableDependencies(expression, getInputVariableDependencies(var.getExpression()));
+                addOutputVariableDependency(expression, var.getVariableName());
+            } else {
+                addInputVariableDependency(expression, var.getVariableName());
+            }
+        }
+
+        removeInputVariableDependencies(
+            expression,
+            getOutputVariableDependencies(expression.getPreviousClause())
+        );
+        return null;
+    }
+
+    public Void visitOrderByClause(OrderByClause expression, Void argument) {
+        visit(expression.getPreviousClause(), null);
+        addOutputVariableDependencies(expression, getOutputVariableDependencies(expression.getPreviousClause()));
+
+        visit(expression.getPreviousClause(), null);
+        for (OrderByClauseSortingKey var : expression.getSortingKeys()) {
+            visit(var.getExpression(), null);
+            addInputVariableDependencies(expression, getInputVariableDependencies(var.getExpression()));
+        }
+
+        removeInputVariableDependencies(
+            expression,
+            getOutputVariableDependencies(expression.getPreviousClause())
+        );
+        return null;
+    }
+
+    public Void visitWhereClause(WhereClause expression, Void argument) {
+        visit(expression.getPreviousClause(), null);
+        addOutputVariableDependencies(expression, getOutputVariableDependencies(expression.getPreviousClause()));
+
+        visit(expression.getWhereExpression(), null);
+        addInputVariableDependencies(expression, getInputVariableDependencies(expression.getWhereExpression()));
+
+        removeInputVariableDependencies(
+            expression,
+            getOutputVariableDependencies(expression.getPreviousClause())
+        );
+        return null;
+    }
+
+    public Void visitCountClause(CountClause expression, Void argument) {
+        visit(expression.getPreviousClause(), null);
+        addOutputVariableDependencies(expression, getOutputVariableDependencies(expression.getPreviousClause()));
+
+        addInputVariableDependencies(expression, new HashSet<String>());
```

Comment:
Does this line do anything? Passing an empty set to addInputVariableDependencies simply returns.

Reply:
Nice catch!